### PR TITLE
Feature/1176/269 jwt audience claim

### PR DESCRIPTION
**Pull request by algattik**
## What this PR changes/adds

Adapted common crypto code to enforce token claim check:
- Audience must match IDS Webhook address
- Subject must be "verifiable-credential"
- Issuer must be set (to any string)
- Expiration time must be set, and in the future

This code is used by the DID authentication module for IDS, but can be used in other contexts as well.

Client generates JWT for IDS requests containing destination web hook HTTP address as audience claim. Server rejects requests with wrong audience.

## Why it does that

Without this change, the server can reuse the token to impersonate the client at a third party

## Further notes

Removed TokenRepresentation.expiredIn, as it was not used, redundant with the token value, and incorrectly set.

Adapted OAuth2 module only to follow interface changes, without implementing additional logic.

## Linked Issue(s)

#269 

## Checklist

- [ ] added appropriate tests?
- [ ] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [ ] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: data-protocols/ids/ids-api-configuration/src/main/java/org/eclipse/dataspaceconnector/ids/api/configuration/IdsApiConfigurationExtension.java [hunk 1]
```diff
@@ -61,8 +65,12 @@ public void initialize(ServiceExtensionContext context) {
         }
         
         monitor.info(format("IDS API will be available at [path=%s], [port=%s].", path, port));
-        
-        context.registerService(IdsApiConfiguration.class, new IdsApiConfiguration(contextAlias, path));
+
+        var idsWebhookAddress = context.getSetting(IDS_WEBHOOK_ADDRESS, DEFAULT_IDS_WEBHOOK_ADDRESS);
+        var webhookPath = path + (path.endsWith("/") ? "data" : "/data");
+        idsWebhookAddress = idsWebhookAddress + webhookPath;
```
Comment (algattik) [inline data-protocols/ids/ids-api-configuration/src/main/java/org/eclipse/dataspaceconnector/ids/api/configuration/IdsApiConfigurationExtension.java:71]: NB: this code was moved from `IdsMultipartDispatcherServiceExtension`

Comment (pkirch) [reply data-protocols/ids/ids-api-configuration/src/main/java/org/eclipse/dataspaceconnector/ids/api/configuration/IdsApiConfigurationExtension.java:71]: nit: while moving this code you could refactor it using the `java.net.URL` class for the path operations.

Comment (algattik) [reply data-protocols/ids/ids-api-configuration/src/main/java/org/eclipse/dataspaceconnector/ids/api/configuration/IdsApiConfigurationExtension.java:71]: That class being "very old" is not very developer friendly, an alternative could be to use OkHttp's URLBuilder class as well

File: extensions/iam/decentralized-identity/identity-did-crypto/src/main/java/org/eclipse/dataspaceconnector/iam/did/crypto/credentials/VerifiableCredentialFactory.java [hunk 2]
```diff
@@ -33,77 +31,45 @@
...
+        var claimsSet = new JWTClaimsSet.Builder().issuer(issuer)
+                .issuer(issuer)
+                .subject(VERIFIABLE_CREDENTIAL)
+                .audience(audience)
                 .expirationTime(Date.from(clock.instant().plus(10, ChronoUnit.MINUTES)))
                 .jwtID(UUID.randomUUID().toString())
                 .build();
 
-        JWSSigner signer = privateKey.signer();
+        var signer = privateKey.signer();
```
Comment (pkirch) [inline extensions/iam/decentralized-identity/identity-did-crypto/src/main/java/org/eclipse/dataspaceconnector/iam/did/crypto/credentials/VerifiableCredentialFactory.java:70]: nit: Could (and imho should) stay `JWSSigner signer` as the type should not change and the method call does not reveal specific type information.

**Review by pkirch — CHANGES_REQUESTED** [review]
Great work! However, it would be easier to review if there would be opportunities to reduce the scope (size) of the PR.

There are several changed code files which would need to add Microsoft to the contributors header.

File: data-protocols/ids/ids-api-multipart-dispatcher-v1/src/test/java/org/eclipse/dataspaceconnector/ids/api/multipart/dispatcher/sender/IdsMultipartSenderTest.java [hunk 3]
```diff
@@ -32,19 +33,21 @@
 import static org.mockito.Mockito.when;
 
 class IdsMultipartSenderTest {
+    private static final Faker FAKER = new Faker();
```
[the remark this answers is not in the document]
Comment (algattik) [reply data-protocols/ids/ids-api-multipart-dispatcher-v1/src/test/java/org/eclipse/dataspaceconnector/ids/api/multipart/dispatcher/sender/IdsMultipartSenderTest.java:36]: Actually a good reason for using Faker is to randomise test inputs to uncover more logical errors.

**Comment by algattik** [comment]
> There are several changed code files which would need to add Microsoft to the contributors header.
thanks, updated

File: extensions/iam/decentralized-identity/identity-did-crypto/src/main/java/org/eclipse/dataspaceconnector/iam/did/crypto/credentials/VerifiableCredentialFactory.java (hunk 2 above)
Comment (pkirch) [reply extensions/iam/decentralized-identity/identity-did-crypto/src/main/java/org/eclipse/dataspaceconnector/iam/did/crypto/credentials/VerifiableCredentialFactory.java:68]: IMHO this is not about consistency. Keywords should be used where they make sense.

However, if this is an EDC coding convention no change required here. 🤷‍♂️

**Review by pkirch — APPROVED** [review]

File: data-protocols/ids/ids-api-configuration/src/main/java/org/eclipse/dataspaceconnector/ids/api/configuration/IdsApiConfigurationExtension.java [hunk 4]
```diff
@@ -61,8 +66,12 @@ public void initialize(ServiceExtensionContext context) {
...
```
Comment (Izzzu) [inline data-protocols/ids/ids-api-configuration/src/main/java/org/eclipse/dataspaceconnector/ids/api/configuration/IdsApiConfigurationExtension.java:72]: nit:
```suggestion
        var idsWebhookAddress = context.getSetting(IDS_WEBHOOK_ADDRESS, DEFAULT_IDS_WEBHOOK_ADDRESS) + webhookPath;
```
and removing line 70

File: extensions/iam/decentralized-identity/identity-did-service/src/main/java/org/eclipse/dataspaceconnector/identity/DecentralizedIdentityService.java [hunk 5]
```diff
@@ -36,36 +38,33 @@
...
         var token = jwt.serialize();
-        var expiration = clock.millis() + TimeUnit.MINUTES.toMillis(10);
-
-        return Result.success(TokenRepresentation.Builder.newInstance().token(token).expiresIn(expiration).build());
+        return Result.success(TokenRepresentation.Builder.newInstance().token(token).build());
     }
 
     @Override
-    public Result<ClaimToken> verifyJwtToken(TokenRepresentation tokenRepresentation) {
+    public Result<ClaimToken> verifyJwtToken(TokenRepresentation tokenRepresentation, String audience) {
```
Comment (Izzzu) [inline extensions/iam/decentralized-identity/identity-did-service/src/main/java/org/eclipse/dataspaceconnector/identity/DecentralizedIdentityService.java:67]: Do we need the audience parameter here if the audience is also a part of the token? 
Is there a case when the audience is not a part of the token?

Comment (algattik) [reply extensions/iam/decentralized-identity/identity-did-service/src/main/java/org/eclipse/dataspaceconnector/identity/DecentralizedIdentityService.java:67]: The audience parameter is the audience we want to check against, ie the value of the audience claim we demand to be in the token

**Review by Izzzu — APPROVED** [review]
Added 2 nit suggestions, looks great!

**Pull request merged**